docs/proposals/thin-volumes.md

```diff
@@ -0,0 +1,189 @@
+
+
+#### Monitoring thin-pools
+- We need to constantly monitor the thin-pools and ensure that it does not run out of space.
+- `/etc/lvm/lvm.conf` settings can allow auto growth of the thin pool when required. By default, the threshold is 100% which means that the pool will not grow.
+- If we set this to, 75%, the Thin Pool will autoextend when the pool is 75% full. It will increase by the default percentage of 20% if the value is not changed. We can see these settings using the command grep against the file.
+
+```go=
+# grep -E ‘^\s*thin_pool_auto’ /etc/lvm/lvm.conf
+ thin_pool_autoextend_threshold = 100
+ thin_pool_autoextend_percent = 20
+```
+- We assume that user has the threshold of 100% with no autogrow.
+- Topolvm will regularly monitor the used thin-pool space for `pool data` and `metadata`.
+- These usages will be exported as metrics so that admin/opeartor can analyse them and take precautionary steps whens usage exeeds certain threshold limit.
+- Command to be used for montoring:
+``` go=
+   lvs
+   LV          VG   Attr       LSize   Pool  Origin Data%  Meta%  Move Log Cpy%Sync Convert
+   lv1         vg1  -wi-ao---- 600.00m
+   thin_volume vg1  Vwi-a-tz--   1.00g tpool        3.20
+   tpool       vg1  twi-aotz-- 200.00m              16.38  1.37
+```
+- Topolvm will only monitor the thin-pools provided in the lvmd.yaml config file.
+
+#### Capacity awareness
+- Topolvm uses capacity awareness to ensure that new pod is created on the node where available capacity of the volume group is larger than the requested capacity.`capacity.topolvm.cybozu.com/<device-class>` annotation is added on each node which consists of the available capacity on the volume group.
+- Creation of thin logical volumes doesn't update the available device class capacity until the user actually uses storage on the thin volumes.
+- So we need to ensure that all the thin logical volumes don't get created on the same node.
+- This can be done by using the `overprovisionRatio`.
+- Value of `capacity.topolvm.cybozu.com/<device-class>_mlog<thin-pool>` annotation will be calculated as:
+  `(thin-pool-size * overprovisionedRatio)-(sum-of-all-thin-lv-sizes-in-thin-pool)`
+- Available thin pools on the volume groups will be obtained from lvmd.yaml config file.
+
+
+
+## Open Questions
+- Should topolvm create the thin-pools instead of the admin/operator?
+
+- If multiple thin-pools are supported for each volume group, then how to decide which thin-pool should be used to create the thin logical volumes.
+    - User should decide which thin-pool to use. This can done via a StorageClass parameter.
+
+- How to do capacity awareness of thin-pools when Volume groups have both thick and thin logical volumes.
+
+- Should we expect auto-extend of thin-pool to be enabled by default? What happens when the user has already enabled the autoextend feature when doing the monitoring?
+
+- Does topolvm constantly poll the available capacity of the thin pools?
```

Comment:
- We would like have an opinion from you about pooling for thin pool size for every minute?
- Is it fine if we do not implement this now and think of other ways if possible?

Reply:
> We would like have an opinion from you about polling for thin pool size for every minute?

It's desirable because it's the simplest solution.

> Is it fine if we do not implement this now and think of other ways if possible?

It depends on the reason. Are there any reasons that you don't want to do so? IMO, it's not so difficult to implement and it can avoid the problem that anyone outside K8s uses thin pools written in device classes.

Reply:
@satoru-takeuchi / @cupnes I was awaiting initial round of reviews on implementation to discuss on pending items however expanding this conversation itself to think about it in parallel.

Problem Statement:
- When the thin pool is expanded by admin topolvm will not realise the increase in size instantly unless an operation is performed which sends `s.notify()`

Solution that's discussed here:
- Use the existing mechanism to get thin pool size and update the node annotation every minute using `checkFunc` and `NewChecker`

Concerns:
- currently annotation and monitoring updates are triggered by `notify` event for all the device-classes
- we can't target a single device-class annotation be updated
> It's desirable because it's the simplest solution.
- by above, you mean the solution is straightforward, if yes, could you pls expand your answer a little bit about how the components should interact to achieve this?

My thoughts:
- we could just document to wait for 10 min after expanding the thin pool to realise the new size as I could see a ticker sending `notify` every 10 min here https://github.com/topolvm/topolvm/blob/f76b553b2d0bdf2ac71d009a4cd0ec36cd9aea5d/pkg/lvmd/cmd/root.go#L112-L123
- pls correct me if above is for something else entirely

Reply:
We will reply next week.

Reply:
@leelavg I'll get back to you later on @cupnes 's behalf because he's extremely busy right now with another jobs.

Reply:
- ack, noted. Meanwhile, a review on implementation PR would also be helpful 😅

Reply:
> I don't think it will be a serious problem. Please let me know if you have any concern.
- now it's all good, what I meant earlier was, this particular function was invoking `getlvbytes` on a default device class to verify lvm subsytem is functional  as a means of readiness checks
- using this functionality to update the node annotation is a little bit contrived

discussion on this thread is finalised, this can be resolved IMHO.

Reply:
@leelavg
[satoru-takeuchi](https://github.com/satoru-takeuchi) is urgent today and I respond this question.

> now it's all good, what I meant earlier was, this particular function was invoking getlvbytes on a default device class to verify lvm subsytem is functional as a means of readiness checks
>
> using this functionality to update the node annotation is a little bit contrived

As you recognize. We don't need to modify around `checkFunc` function. We had to only modify `vgService.send` to update node labels of pool capacity. Could you please update the answer.

- TopoLVM updates the node's annotation every 10 mins using `notify` function as a trigger as you know.
- The annotations are updated on `metricsExporter.updateNode` by values get using `VGServece.Watch`.
- Modify `vgService.send` to update value used by `VGServece.Watch`.

> discussion on this thread is finalised

I agree with you.

Reply:
> Modify vgService.send to update value used by VGServece.Watch.
- this is already part of the implementation, I'll add how metrics & annotations are being updated to this design doc.